src/datajunction/superset.py

```diff
@@ -0,0 +1,120 @@
+"""
+A DB engine spec for Superset.
+"""
+
+import re
+from typing import TYPE_CHECKING, Any, List, Optional, TypedDict
+
+import requests
+from sqlalchemy.engine.reflection import Inspector
+
+try:
+    from superset.db_engine_specs.base import BaseEngineSpec
+except ImportError:  # pragma: no cover
+    # we don't really need the base class, so we can just mock it if Apache Superset is
+    # not installed
+    BaseEngineSpec = object
+
+if TYPE_CHECKING:
+    from superset.models.core import Database
+
+
+SELECT_STAR_MESSAGE = (
+    "DJ does not support data preview, since the `metrics` table is a virtual table "
+    "representing the whole repository of metrics. An administrator should configure the "
+    "DJ database with the `disable_data_preview` attribute set to `true` in the `extra` "
+    "field."
+)
+
+
+class MetricType(TypedDict, total=False):
+    """
+    Type for metrics return by `get_metrics`.
+    """
+
+    metric_name: str
+    expression: str
+    verbose_name: Optional[str]
+    metric_type: Optional[str]
+    description: Optional[str]
+    d3format: Optional[str]
+    warning_text: Optional[str]
+    extra: Optional[str]
+
+
+class DJEngineSpec(BaseEngineSpec):  # pylint: disable=abstract-method
+    """
+    Engine spec for the DataJunction metric repository
+
+    See https://github.com/DataJunction/datajunction for more information.
+    """
+
+    engine = "dj"
+    engine_name = "DJ"
+
+    sqlalchemy_uri_placeholder = "dj://host:port/database_id"
+
+    _time_grain_expressions = {
+        None: "{col}",
+        "PT1S": "DATE_TRUNC('second', {col})",
+        "PT1M": "DATE_TRUNC('minute', {col})",
+        "PT1H": "DATE_TRUNC('hour', {col})",
+        "P1D": "DATE_TRUNC('day', {col})",
+        "P1W": "DATE_TRUNC('week', {col})",
+        "P1M": "DATE_TRUNC('month', {col})",
+        "P3M": "DATE_TRUNC('quarter', {col})",
+        "P1Y": "DATE_TRUNC('year', {col})",
+    }
+
+    @classmethod
+    def select_star(  # pylint: disable=unused-argument
+        cls,
+        *args: Any,
+        **kwargs: Any,
+    ) -> str:
+        """
+        Return a ``SELECT *`` query.
+
+        Since DJ doesn't have tables per se, a ``SELECT *`` query doesn't make sense.
+        """
+        message = SELECT_STAR_MESSAGE.replace("'", "''")
+        return f"SELECT '{message}' AS warning"
+
+    @classmethod
+    def get_metrics(  # pylint: disable=unused-argument
+        cls,
+        database: "Database",
```

Comment:
What does this type mean `"Database"` ?

Reply:
This a Superset `Database` object... we only use it for type annotation, so instead of importing the actual `Database` class and using it here (which would require to have Superset installed) we can just reference it as a string and do the import under a `if TYPE_CHECKING` check:

https://github.com/DataJunction/datajunction/pull/175/files#diff-4ef254f7258961794702764475cd8683b37116cf4c0c24d803c76907556fe945R18-R19.

It's a little bit clunky. But usually when you have cyclic dependencies you use the type as a string. For example, this will fail:

```python
class Parent:
    child: Child  # NameError: name 'Child' is not defined

class Child:
    parent: Parent
```

So we do:

```python
class Parent:
    child: "Child"

class Child:
    parent: "Parent"
```

There are better ways of doing this, starting in Python 3.7:

```python
from __future__ import annotations

class Parent:
    child: Child  # Now it works!

class Child:
    parent: Parent
```

This will be the standard in 3.11. Since DJ supports 3.8+, maybe we should add the `from __future__ import annotations` to relevant files.